Shared/mods/deathmatch/logic/luadefs/CLuaCryptDefs.cpp

```diff
@@ -77,100 +78,75 @@ std::string CLuaCryptDefs::Base64decode(std::string str)
-                            },
-                            [luaFunctionRef](const SString& hash) {
-                                CLuaMain* pLuaMain = m_pLuaManager->GetVirtualMachine(luaFunctionRef.GetLuaVM());
-                                if (pLuaMain)
-                                {
-                                    CLuaArguments arguments;
-
-                                    if (hash.empty())
-                                    {
-                                        m_pScriptDebugging->LogCustom(pLuaMain->GetVM(), "Invalid value for field 'salt'");
-                                        arguments.PushBoolean(false);
-                                    }
-                                    else
-                                        arguments.PushString(hash);
+                    CLuaShared::GetAsyncTaskScheduler()->PushTask<SString>(
+                        [password, salt = options["salt"], cost] {
+                            // Execute time-consuming task
+                            return SharedUtil::BcryptHash(password, salt, cost);
+                        },
+                        [luaFunctionRef = callback.value()](const SString& hash) {
+                            CLuaMain* pLuaMain = m_pLuaManager->GetVirtualMachine(luaFunctionRef.GetLuaVM());
+                            if (pLuaMain)
+                            {
+                                CLuaArguments arguments;
 
-                                    arguments.Call(pLuaMain, luaFunctionRef);
+                                if (hash.empty())
+                                {
+                                    m_pScriptDebugging->LogCustom(pLuaMain->GetVM(), "Invalid value for field 'salt'");
+                                    arguments.PushBoolean(false);
                                 }
-                            });
+                                else
+                                    arguments.PushString(hash);
 
-                        lua_pushboolean(luaVM, true);
-                        return 1;
-                    }
+                                arguments.Call(pLuaMain, luaFunctionRef);
+                            }
+                        });
                 }
             }
             else
-                m_pScriptDebugging->LogWarning(luaVM, "Invalid value for field 'cost'");
+            {
+                // Sync
+                SString hash = SharedUtil::BcryptHash(password, options["salt"], cost);
+                if (!hash.empty())
+                {
+                    return hash;
+                }
+                else
+                    throw std::invalid_argument("Invalid value for field 'salt'");
+            }
+            return true;
         }
-    }
-    else
-        m_pScriptDebugging->LogCustom(luaVM, argStream.GetFullErrorMessage());
 
-    lua_pushboolean(luaVM, false);
-    return 1;
+        default:
+            throw std::invalid_argument("Unknown algorithm");
+    }
 }
```

Comment:
For some reason cost nor salt option is working for me (occurs for both sync and async) - am I using this function wrong? Same result on client and server-side.

Cost (server-side):
```
[2021-03-21 22:34:31] Console executed command: passwordHash("password", "bcrypt", { cost = 10 })
[2021-03-21 22:34:31] WARNING: [admin]\runcode\server.lua:21: Invalid value for field 'cost' [string "return passwordHash("password", "bcrypt", {..."]
[2021-03-21 22:34:31] Command results: false [boolean]
```

Salt (client-side):
```
[2021-03-21 22:38:46] [Input]  : crun passwordHash("password", "bcrypt", { salt = "abcdefg" })
[2021-03-21 22:38:46] [Output] : Executing client-side command: passwordHash("password", "bcrypt", { salt = "abcdefg" })
[2021-03-21 22:38:46] WARNING: runcode\client.lua:19: Custom generated salts are deprecated and will be removed in the future. [string "return passwordHash("password", "bcrypt", {..."]
[2021-03-21 22:38:46] WARNING: runcode\client.lua:19: Invalid value for field 'salt' [string "return passwordHash("password", "bcrypt", {..."]
[2021-03-21 22:38:46] [Output] : Command results: false [boolean]
```

Reply:
I fixed the cost issue.
The salt issue is on your side, as it has to be >= 22 in length.